### ISSUE TITLE
Memory hotplug interacts badly with NVDIMM

### ISSUE THREAD

**Issue by rbradford** [intel/nemu#48]
Memory hotplug does not work with OVMF ([0.2](https://github.com/rbradford/edk2/releases/tag/0.2))

Testing with Clear Linux https://download.clearlinux.org/releases/24380/clear/clear-24380-cloud.img.xz (has @sboeuf kernel changes)

When hotplugging 1G dimm:

```
(qemu) object_add memory-backend-ram,id=mem1,size=1G
(qemu) device_add pc-dimm,id=dimm1,memdev=mem1
```

I see the following in dmesg:

```
[  676.541766] Block size [0x8000000] unaligned hotplug range: start 0x100200000, size 0x40000000
[  676.541770] acpi PNP0C80:01: add_memory failed
[  676.542916] acpi PNP0C80:01: acpi_memory_enable_device() error
[  676.543412] acpi PNP0C80:01: Enumeration failure
```
CC: @sboeuf @yangzhon @sameo 

qemu command line is: 
```
/home/rob/build-x86_64/x86_64-softmmu/qemu-system-x86_64 -machine virt,accel=kvm,kernel_irqchip,nvdimm -pidfile qemu.pid -monitor telnet:127.0.0.1:55555,server,nowait -bios /home/rob/src/edk2/Build/OvmfX64/DEBUG_GCC5/FV/OVMF.fd -smp 2,cores=1,threads=1,sockets=2,maxcpus=32 -m 512,slots=4,maxmem=16384M -cpu host -nographic -no-user-config -nodefaults -daemonize -drive file=testvm.img,if=none,id=drive-virtio-disk0,format=qcow2 -device virtio-blk-pci,scsi=off,drive=drive-virtio-disk0,id=virtio-disk0,bootindex=1 -device sysbus-debugcon,iobase=0x402,chardev=debugcon -chardev file,path=/tmp/debug-log,id=debugcon -device virtio-blk-pci,drive=cloud -drive if=none,id=cloud,file=seed.img,format=raw -netdev user,id=mynet0,hostfwd=tcp::2222-:22,hostname=nemuvm -device virtio-net-pci,netdev=mynet0 -drive file=testscsi.img,if=none,id=drive-virtio-disk1,format=raw -device virtio-scsi-pci,id=virtio-disk1 -object memory-backend-file,id=mem0,share,mem-path=testnvdimm.img,size=51200 -device nvdimm,memdev=mem0,id=nv0 -device virtio-serial-pci,id=virtio-serial0 -device virtconsole,chardev=charconsole0,id=console0 -chardev socket,id=charconsole0,path=console.sock,server,nowait -netdev tap,fd=3,id=hostnet0,vhost=on,vhostfd=4 -device virtio-net-pci,netdev=hostnet0,id=net0,mac=1a:a2:11:c5:a5:11 -device virtio-rng-pci,rng=rng0 -object rng-random,filename=/dev/random,id=rng0 -device virtio-balloon-pci -object cryptodev-backend-builtin,id=cryptodev0 -device virtio-crypto-pci,id=crypto0,cryptodev=cryptodev0
```

**Comment by liujing2** [comment]
I just tested this on direct kernel boot, with config including https://github.com/intel/nemu/pull/41#issuecomment-411875634, and clear-24380-cloud.img, getting similar result.
```
acpi PNP0C80:00: Enumeration failure
```

**Comment by rbradford** [comment]
Removing the added NVDIMM device makes hotplug of memory work:

i.e. removing this:

`-object memory-backend-file,id=mem0,share,mem-path=testnvdimm.img,size=51200 -device nvdimm,memdev=mem0,id=nv0`

**Comment by rbradford** [comment]
Also, adjusting the size of the nvdimm to 1Gib (from 50M) makes no difference:

```
[   38.020112] Block size [0x8000000] unaligned hotplug range: start 0x100200000, size 0x40000000
[   38.020124] acpi PNP0C80:01: add_memory failed
[   38.028180] acpi PNP0C80:01: acpi_memory_enable_device() error
[   38.030116] acpi PNP0C80:01: Enumeration failure
```

**Comment by yangzhon** [comment]
@rbradford @sboeuf 

as for log:
unaligned hotplug range: start 0x100200000

The hotplug base address should be 0x100000000

**Comment by rbradford** [comment]
FYI, with Q35 and upstream OVMF.fd I see the same error message and the memory does not get added to MemTotal

**Comment by rbradford** [comment]
It also fails if I use qemu 2.12.0 (so none of our NEMU changes) and QEMU tip

**Comment by rbradford** [comment]
So the problem is this:

When you have the nvdimm attached the next address for the memory hotplug may not be aligned to  128mib hence it will fail.

If you change your nvdimm to something that is a multiple of 128mib then it correctly works. The start_qemu.sh script had two bugs:

* It was reporting the size in K and treating it as bytes (so my attempt to give me 1gib nvdimm gave me a  1mb (not mib)
* It needs to use an nvdimm device of a multiple of 128mib

**Comment by sameo** [comment]
@rbradford I feel this is a qemu upstream bug. For sure the memory hotplug range should be aligned on 2^27, and qemu should be able to pick the next available one.

**Comment by sameo** [comment]
@rbradford This should be fixed with PR #50 ?

**Comment by sboeuf** [comment]
Yes it should be
